File: java/strings/easy/ValidParentheses.java
class Solution {
    public static final Map<Character, Character> map = Map.of('(',')',
                                                               '{','}',
                                                               '[',']');
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        for(char c : s.toCharArray()) {
            if(map.containsKey(c)) {
                stack.push(c);
            } else {
                if(stack.isEmpty()) {
                    return false;
                }
                char open = stack.pop();
                if(c != map.get(open)) {
                    return false;
                }
            }
        }
        
        return stack.isEmpty();
    }
}

File: Python/Strings/Easy/validParentheses.py
class Solution(object):
    def validParentheses(self, s):
        # init stack data structure
        stack = []

        # init mappings for valid parentheses
        mappings = {')':'(',
                    ']':'[',
                    '}':'{' }

        # loop through chars in string
        for char in s:

            # check if in mapping 
            if char in mappings:
                # if match found in mapping, pop top most element 
                top_element = stack.pop() if stack else '#'

                if mappings[char] != top_element:
                    return False 
            else:
                stack.append(char)
        return not stack 

#Time comp: O(n)
#Space comp: O(n)